lib/teleterm/daemon/gateway_cert_reissuer.go

```diff
@@ -0,0 +1,185 @@
+// notification. GatewayCertReissuer is typically called from within a goroutine that handles the
+// gateway, so without forwarding the error to the app, it would be visible only in the logs.
+func (r *GatewayCertReissuer) ReissueCert(ctx context.Context, gateway *gateway.Gateway, dbCertReissuer DBCertReissuer) error {
+	err := r.reissueCert(ctx, gateway, dbCertReissuer)
+	if err != nil {
+		r.notifyAppAboutError(ctx, err, gateway)
+
+		// Return the error to the alpn.LocalProxy's middleware.
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (r *GatewayCertReissuer) reissueCert(ctx context.Context, gateway *gateway.Gateway, dbCertReissuer DBCertReissuer) error {
+	// Make the first attempt at reissuing the db cert.
+	//
+	// It might happen that the db cert has expired but the user cert is still active, allowing us to
+	// obtain a new db cert without having to relogin first.
+	//
+	// This can happen if the user cert was refreshed by anything other than the gateway itself. For
+	// example, if you execute `tsh ssh` within Connect after your user cert expires or there are two
+	// gateways that subsequently go through this flow.
+	err := r.reissueAndReloadGatewayCert(ctx, gateway, dbCertReissuer)
+
+	if err == nil {
+		return nil
+	}
+
+	// Do not ask for relogin if the error cannot be resolved with relogin.
+	if !client.IsErrorResolvableWithRelogin(err) {
+		return trace.Wrap(err)
+	}
+
+	clusterURI, err := uri.ParseClusterURI(gateway.TargetURI())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	rootClusterURI := clusterURI.GetRootClusterURI().String()
+
+	err = r.requestReloginFromElectronApp(ctx,
+		&api.ReloginRequest{
+			RootClusterUri: rootClusterURI,
+			Reason: &api.ReloginRequest_GatewayCertExpired{
+				GatewayCertExpired: &api.GatewayCertExpired{
+					GatewayUri: gateway.URI().String(),
+					TargetUri:  gateway.TargetURI(),
+				},
+			},
+		})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = r.reissueAndReloadGatewayCert(ctx, gateway, dbCertReissuer)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (r *GatewayCertReissuer) reissueAndReloadGatewayCert(ctx context.Context, gateway *gateway.Gateway, dbCertReissuer DBCertReissuer) error {
+	err := dbCertReissuer.ReissueDBCerts(ctx, gateway.RouteToDatabase())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(gateway.ReloadCert())
+}
+
+func (r *GatewayCertReissuer) requestReloginFromElectronApp(ctx context.Context, req *api.ReloginRequest) error {
+	const reloginUserTimeout = time.Minute
+	timeoutCtx, cancelTshdEventsCtx := context.WithTimeout(ctx, reloginUserTimeout)
+	defer cancelTshdEventsCtx()
+
+	// The Electron app cannot display two login modals at the same time, so we have to cut short any
+	// concurrent relogin requests.
+	if !r.reloginMu.TryLock() {
+		return trace.AlreadyExists("another relogin request is in progress")
+	}
+	defer r.reloginMu.Unlock()
+
+	_, err := r.TSHDEventsClient.Relogin(timeoutCtx, req)
```

Comment:
@GavinFrazar Are we able to detect if the program connecting to the proxy has dropped the connection?

Right now [the context that is passed to `OnNewConnection`  is tied solely to the lifecycle of LocalProxy itself](https://github.com/gravitational/teleport/blob/59a96be624ede92e6dcddaf6cf62c33bbe85de98/lib/srv/alpnproxy/local_proxy.go#L142-L170). This means that if this code here waits for the Electron app to relogin the user and the db client drops the connection, this call won't get aborted and the Electron app will still show a login modal.

I did a brief search and it seems that we could have [a separate goroutine that peeks at the connection](https://stackoverflow.com/a/58664631/742872). Although tbh this is all a bit over my head. If it's not something we can easily implement then I guess we have to live with it for now.

Reply:
What happens if the user logs in via modal after the db connect is dropped?

I can see how it would be frustrating to log in for no reason or to immediately get an error.

The problem with any connection drop detection, is that you may (somehow) check if the connection has dropped, but it hasn't yet, then you proceed and the connection is dropped after your check. I think I've tried to do this before for some other issue I was investigating, and the conclusion I came to was basically that you have to follow the idiom "don't look before you leap" (a python saying), in other words it's better to try and fail than to establish some prerequisite which may no longer be true once you "leap". So unfortunately, I don't see a way to avoid this problem.

Reply:
> What happens if the user logs in via modal after the db connect is dropped?

Nothing bad is going to happen, the user cert and the db cert will get refreshed and the gateway will use the new db cert. When the user tries to connect again, the connection should be successfully established.

Unfortunate that we can't address this but yeah, it's not the end of the world.